### ISSUE TITLE
help request: How to subscribe to and receive kafka through Apisix

### ISSUE THREAD

**Issue by EmptyNest111** [apache/apisix#10421]
### Description

May I ask if I found during the process of registering Kafka using Apisix that the registered ws:/127.0.0.1:9080/kafka can be linked using Postman's websocket test, but the official website does not specify how to receive data sent by producers, nor does it specify which topic to subscribe to....

### Environment

- APISIX version (run `apisix version`): 2.15.0
- Operating system (run `uname -a`): Linux hibasebrain4test-worker-3 3.10.0-957.el7.x86_64
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`): 1.21.4.2
- etcd version, if relevant (run `curl http://127.0.0.1:9090/v1/server_info`): 3.5.0
- APISIX Dashboard version, if relevant:
- Plugin runner version, for issues related to plugin runners:
- LuaRocks version, for installation issues (run `luarocks --version`): 3.8.0

**Comment by Revolyssup** [comment]
@EmptyNest111 Can you paste the relevant route(or other) configuration that you're using here? 
Are you trying to use kafka-proxy plugin?

**Comment by shreemaan-abhishek** [comment]
This might help you: https://apisix.apache.org/docs/apisix/pubsub/kafka/

**Comment by EmptyNest111** [comment]
> @EmptyNest111 Can you paste the relevant route(or other) configuration that you're using here? Are you trying to use kafka-proxy plugin?

yes,I want to register the scheme in Upstream as kafka, and then test using the registered URL to obtain messages,I have checked the content of the pubsub module on the official website, but I feel it is very complex and not very clear. Is it necessary to customize the development of Apisix?

**Comment by EmptyNest111** [comment]
> This might help you: https://apisix.apache.org/docs/apisix/pubsub/kafka/



>I didn't understand this pubsub. I just wanted to test the feasibility of receiving subscription messages. Does it require customized development of Apisix to be implemented?

**Comment by monkeyDluffy6017** [comment]
@EmptyNest111 have you solved your problem?

**Comment by dannyboodmann** [comment]
Also for me it's not very clear how to proceed when one wants to setup APISIX in order to collect information from Kafka. I've configured my route following the documentation... now what? :-)

From the doc:

> "First, it is necessary to compile the [communication protocol](https://github.com/apache/apisix/blob/master/apisix/include/apisix/model/pubsub.proto) as a language-specific SDK using the protoc, which provides the command and response definitions to connect to Kafka via APISIX using the WebSocket."

If I understand this correctly, this means that the protocol needs to be compiled in the language we want to use. For example, for Java one would follow something like this: https://protobuf.dev/getting-started/javatutorial/ . Do I get it right?

Also, it looks like to me that it's "only" possible to _pull_ messages from Kafka using APISIX, not _pushing_ messages to it. Is my understanding correct?

Thanks in advance for the replies,

Marco

**Comment by senolatac** [comment]
@EmptyNest111 did you find a solution?

**Comment by kayx23** [comment]
> Also, it looks like to me that it's "only" possible to pull messages from Kafka using APISIX, not pushing messages to it. Is my understanding correct?

@dannyboodmann correct

**Comment by kayx23** [comment]
Disclaimer: I'm no expert on this topic, but here are some information I gathered.

The test case (https://github.com/apache/apisix/blob/master/t/pubsub/kafka.t#L156) suggests that one should pull messages from kafka over websocket connections and with messages binary-encoded.

Only `list_offset` and `fetch` commands are supported at the moment. `list_offset` is used to check the message offset at the specified timestamp on the current topic. `fetch` is used to obtain the message from a specific offset.
